### PR TITLE
YOLOv3 loss function

### PR DESCRIPTION
**Pull request by andrewfurey21**
## YOLOv3 Loss Function

A loss function that tries to match closely with the darknet repo.

### CODE REVIEW

File: src/mlpack/core/util/using.hpp [hunk 1]
```diff
@@ -91,6 +91,7 @@ using arma::pow;
 using arma::randi;
 using arma::randn;
 using arma::randu;
+using arma::regspace;
```
Comment (rcurtin) [inline src/mlpack/core/util/using.hpp:95]: We should probably add a `using coot::regspace` to the appropriate location down below (or up above? not sure) too.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+template<typename MatType = arma::mat>
+class YOLOv3Loss
+{
+ public:
+  using ElemType = typename MatType::elem_type;
+  using CubeType = typename GetCubeType<MatType>::type;
+  using IndexType = typename GetUColType<MatType>::type;
+
+  YOLOv3Loss() { std::cout << "Default yolo loss.\n"; }
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp]: Probably this is debugging output that can be removed now?

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  /**
+   */
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp:29]: When you have a chance, can you go through and fill these comments out?  It is helpful for users and also maintainers who might come back to this someday 😄  (also a block comment for the class would be helpful too)

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  void Backward(const MatType& predictions,
+                const MatType& targets,
+                const MatType& bestPredictionIndices,
+                const MatType& ignorePredictions,
+                const MatType& scales,
+                const MatType& numTargets,
+                MatType& loss);
+
+  //! Serialize the EmptyLossType.
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp]: Total side note: the `//!` was for Doxygen, but, we haven't used Doxygen in years, so any `//!` instead of `//` are purely vestigial.  I tend to just remove them when I see them.  Anyway, no need to add them to new code.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+ * @file methods/ann/loss_functions/yolov3_loss_fn_impl.hpp
+ * @author Andrew Furey
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_MODELS_YOLOV3_LOSS_FN
+#define MLPACK_METHODS_ANN_MODELS_YOLOV3_LOSS_FN
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp]:
```suggestion
#ifndef MLPACK_METHODS_ANN_MODELS_YOLOV3_LOSS_FN_IMPL_HPP
#define MLPACK_METHODS_ANN_MODELS_YOLOV3_LOSS_FN_IMPL_HPP
```

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    error.cols(conv_to<IndexType>::from(bestPredictionIndices.col(i))) =
+      error.cols(conv_to<IndexType>::from(bestPredictionIndices.col(i))) %
+      (1. - keepObject) + errorTruths % keepObject;
+
+    // at end, ignore boxes as needed, then accumulate into loss scalar.
+    loss +=
+      accu(error % (1. - repmat(ignorePredictions.col(i).t(), numAttributes, 1)));
+  }
+
+  return loss / batchSize;
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp]: If ensmallen optimizers are already going to normalize the gradient to the batch size, then do we need to do this here?  I'm not 100% sure in my comment, but, I think it's worth digging in to figure out what is the right thing to do.  Intuitively I expect that each layer's backwards pass computes the gradient *sum* across the minibatch; then, ensmallen divides that by the batch size and applies it during the SGD-like optimizer's step.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+#ifndef MLPACK_METHODS_ANN_MODELS_YOLOV3_LOSS_FN
+#define MLPACK_METHODS_ANN_MODELS_YOLOV3_LOSS_FN
+
+#include "yolov3_loss_fn.hpp"
+
+namespace mlpack {
+
+template<typename MatType>
+typename YOLOv3Loss<MatType>::ElemType
+YOLOv3Loss<MatType>::Forward(const MatType& predictions,
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:19]: Do you think we can add a test making sure that the forward/backward implementations match?  You can probably base this on `JacobianPerformanceTest()` in `src/mlpack/tests/ann/ann_test_tools.hpp`.  Also maybe a simple hand-coded forward/backward test would be good too (maybe it would be a bit tedious because the inputs and outputs are so large).

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+    // squared error for box coords x, y, w, y.
+    errorTruths.rows(0, 3) = repmat(scales.col(i).t(), 4, 1) %
+      pow(targetsCube.slice(i).rows(0, 3) - correctBoxes.rows(0, 3), 2);
+
+    // binary cross entropy for objectness and classification for boxes
+    // whose anchors match the ground truths the best.
+    probabilities = 1. / (1. + exp(-correctBoxes.rows(4, numAttributes - 1)));
+
+    // Objectness
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:66]: I think you already have this written in other places, but it might be beneficial when you make `predictionsCube` to describe what each actual column/row represents; at least, that would make this part of the code a bit easier to understand.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  for (size_t i = 0; i < batchSize; i++)
+  {
+    error.fill(0);
+
+    // boxes that don't match the targets only affect the objectness loss (objectness should be 0).
+    error.row(4) = -log(1. - (1. / (1. + exp(-predictionsCube.slice(i).row(4)))));
+
+    // gather boxes that match the ground truths
+    correctBoxes = predictionsCube.slice(i).cols(
+      conv_to<IndexType>::from(bestPredictionIndices.col(i)));
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp]: Might be better to convert `bestPredictionIndices` to `IndexType` back when it was created, instead of here (where it will be done every single time, sometimes involving a temporary matrix if it's Armadillo instead of Bandicoot).  It might also be nice to see if `correctBoxes` can be avoided altogether by inlining the expression... but... that is going to make some complex expressions, and I don't know if runtime will benefit much.  So, maybe that bit of optimization should be saved for later when we're actually trying to train.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+      (1 - targetsCube.slice(i).rows(5, numAttributes - 1)) * (probabilities.rows(1, numAttributes - 5));
+
+    keepObject = conv_to<MatType>::from(
+      keepObjectRange < repmat(repeatedNumTargets.col(i), 1, numTruths));
+    lossCube.slice(i).cols(conv_to<IndexType>::from(bestPredictionIndices.col(i))) =
+      lossCube.slice(i).cols(conv_to<IndexType>::from(bestPredictionIndices.col(i))) % (1 - keepObject) +
+      (dLossdBoxes % keepObject);
+
+    lossCube.slice(i) %=
+      (1. - repmat(ignorePredictions.col(i).t(), numAttributes, 1));
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:146]: I am only seeing `ignorePredictions` used with a transpose here, so does it make sense to transpose it when it's created?

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+  MakeAlias(lossCube, loss, numAttributes, numBoxes, batchSize);
+
+  MatType dLossdBoxes = MatType(numAttributes, numTruths);
+
+  MatType probabilities, correctBoxes, keepObject;
+
+  MatType repeatedNumTargets = repmat(numTargets, numAttributes, 1);
+
+  lossCube.row(4) = 1. / (1. + exp(-predictionsCube.row(4)));
+  for (size_t i = 0; i < batchSize; i++)
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:126]: Do you think we could batch this over cubes?  I think you could, with `repcube()` and `.subcube()`, `.tube()`, etc.  That's not likely to make too much difference with Armadillo (although it might because Armadillo may use OpenMP to parallelize), but I do think it will make a noticeable difference with Bandicoot.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+  CubeType predictionsCube, targetsCube, lossCube;
+  MakeAlias(predictionsCube, predictions, numAttributes, numBoxes, batchSize);
+  MakeAlias(targetsCube, targets, numAttributes, numTruths, batchSize);
+  MakeAlias(lossCube, loss, numAttributes, numBoxes, batchSize);
+
+  MatType dLossdBoxes = MatType(numAttributes, numTruths);
+
+  MatType probabilities, correctBoxes, keepObject;
+
+  MatType repeatedNumTargets = repmat(numTargets, numAttributes, 1);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:123]: For Bandicoot at least, this will be much better to just inline.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+  CheckShapes(predictions,
+              targets,
+              bestPredictionIndices,
+              ignorePredictions,
+              scales,
+              numTargets);
+
+  loss.set_size(predictions.n_rows, predictions.n_cols);
+  loss.fill(0);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp]:
```suggestion
  loss.zeros(predictions.n_rows, predictions.n_cols);
```

But, shouldn't the enclosing `DAGNetwork` initialize `loss` to the right size beforehand?

**Review by rcurtin — COMMENTED** [review]
Nice, it looks good!  I have some small suggestions that will maybe be helpful.  I can't say whether or not the implementation is mathematically correct; a couple tests might help out (and we should definitely add those so we have some belief that haven't broken the implementation when making changes in the future).

But, the real test is just whether or not we can train and the loss decreases.  Do you plan to test that before this PR is merged?  Or do you have another PR in mind that will add some changes that will ensure convergence?

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp (hunk 3 above)
Comment (andrewfurey21) [reply src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn.hpp:29]: Yeah, I'll do this last.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp (hunk 11 above)
Comment (andrewfurey21) [reply src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:126]: I'm not sure how to, the main problem being indexing with the `bestPredictionIndices` matrix.  When I do `cube.cols(matrix)` it doesn't compile, it would be nice if you could do that, where each column of the matrix gathers from each slice.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp (hunk 13 above)
Comment (andrewfurey21) [reply src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp]: Actually, it probably should but it doesn't, since it's not part of the forward activations/backward gradients.

Comment (andrewfurey21) [reply src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp]: I'll make a different PR for that.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp (hunk 10 above)
Comment (andrewfurey21) [reply src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:146]: The reason I did that was just to keep n_cols as batchSize. Does it affect performance at all?

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp (hunk 11 above)
Comment (rcurtin) [reply src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:126]: You're right, sorry about that, `Cube`s have no way to indirect addressing of columns.  So... yeah, I guess there's not currently much better option than what we have.

File: src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp (hunk 10 above)
Comment (rcurtin) [reply src/mlpack/methods/ann/models/yolov3/yolov3_loss_fn_impl.hpp:146]: What I overlooked here is that `ignorePredictions.col(i)` is being transposed, not `ignorePredictions`... this amounts to just a logical reshaping of a column vector to a row vector.  So, yeah, we can ignore this comment, sorry about that.
